modules/auxiliary/gather/ldap_query.rb

```diff
@@ -184,6 +184,101 @@ def generate_rex_tables(entries, format)
     end
   end
 
+  # Decode a binary string containing a GUID into a GUID string
+  # of the format XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX
+  def decode_guid_binstring(guid)
+    # See https://docs.microsoft.com/en-us/dotnet/api/system.guid.tobytearray?view=netcore-3.1#remarks
+    # for more information. Specifically note that they state that the first four byte group and the
+    # following 2 groups of two bytes each are reversed. Following this, the next 2 byte group and
+    # the terminating 4 byte group are kept in the same order as in the binary data.
+    #
+    # Thanks to Spencer McIntyre for providing assistance on this code.
+    decoded_guid = [guid[0...4].reverse, guid[4...6].reverse, guid[6...8].reverse, guid[8...10], guid[10..]].map { |p| p.unpack1('H*') }.join('-')
+    decoded_guid = decoded_guid.upcase # All GUID strings must be uppercase
+
+    decoded_guid
+  end
+
+  def convert_nt_timestamp_to_time_string(nt_timestamp)
+    Time.at((nt_timestamp.to_i - 116444736000000000) / 10000000).utc.to_s
+  end
+
+  def convert_pwd_age_to_time_string(timestamp)
+    seconds = (timestamp.to_i / -1) / 10000000 # Convert always negative number to positive then convert to seconds from tick count.
+    days = seconds / 86400
+    hours = (seconds % 86400) / 3600
+    minutes = ((seconds % 86400) % 3600) / 60
+    real_seconds = (((seconds % 86400) % 3600) % 60)
+    return "#{days}:#{hours.to_s.rjust(2, '0')}:#{minutes.to_s.rjust(2, '0')}:#{real_seconds.to_s.rjust(2, '0')}"
+  end
+
+  # Read in a DER formatted certificate file and transform it into a
+  # OpenSSL::X509::Certificate object before then using that object to
+  # read the properties of the certificate and return this info as a string.
+  def read_der_certificate_file(cert)
+    openssl_certificate = OpenSSL::X509::Certificate.new(cert)
+    version = openssl_certificate.version
+    subject = openssl_certificate.subject
+    issuer = openssl_certificate.issuer
+    algorithm = openssl_certificate.signature_algorithm
+    extensions = openssl_certificate.extensions.join(' | ')
+    extensions.strip!
+    extensions.gsub!(/ \|$/, '') # Strip whitespace and then strip trailing | from end of string.
+    [openssl_certificate, "Version: 0x#{version}, Subject: #{subject}, Issuer: #{issuer}, Signature Algorithm: #{algorithm}, Extensions: #{extensions}"]
+  end
+
+  def decode_binary_data_to_sid(object_sid_raw)
+    revision_level = object_sid_raw[0].unpack('H*')[0].to_i.to_s
+    count_subauthorities = object_sid_raw[1].unpack('H*')[0].to_i
+    identifier_authority = object_sid_raw[2..7].unpack('H*')[0].to_i.to_s
+    sid_string = "S-#{revision_level.to_i}-#{identifier_authority.to_i}"
+
+    for i in 0...count_subauthorities
+      start = 8 + (4 * i)
+      stop = start + 4
+      sub_authority_raw = object_sid_raw[start...stop]
+      sub_authority = sub_authority_raw.unpack('L<')[0].to_i.to_s
+      sid_string += "-#{sub_authority}"
+    end
+
+    sid_string
+  end
```

Comment:
Should probably update this to using BinData in some way if we can vs custom unpacks. I think we have this already in other PRs so it may not even be needed at this point.

Reply:
Fixed with https://github.com/rapid7/metasploit-framework/pull/16979/commits/51829c5ce96accb59764cec7e7b330ed03868972